### ISSUE TITLE
Error trying to parse large JSON

### ISSUE THREAD

**Issue by Javier-Rotelli** [mbloch/mapshaper#129]
Hi! i'm trying to simplify this GeoJson https://github.com/jgoodall/us-maps/tree/master/geojson.
the zcta5.json is 1.4g 

mapshaper errors out whenever i try to parse it. i think it's the same error that was reported in #47 namely "Node won't assign a buffer greater than 1gb"

**Comment by mbloch** [comment]
Mapshaper's command line tool could probably handler larger GeoJSON files by switching to a streaming GeoJSON parser. It's worth looking into. Meanwhile, I'd recommend that you work with Shapefile sources. You can download ZCTA5 boundaries from the Census department. Here's a link to the zcta5 dataset from the "Cartographic Boundary Shapefiles" collection:
https://www.census.gov/geo/maps-data/data/cbf/cbf_zcta.html

You can find another version with additional data fields in "TIGER/Line" (https://www.census.gov/geo/maps-data/data/tiger-line.html)

**Comment by Javier-Rotelli** [comment]
woah! that was fast =D. 
yup, i ended going that route, with the shapefile everything went smooth.
But i tought it was something worth reporting. 
if you want i could probably make a PR with a streaming parser.

**Comment by mbloch** [comment]
I see there's a similar issue open here: https://github.com/mbostock/topojson/issues/80
And a streaming GeoJSON parser that's used by a number of other projects here: https://github.com/mapbox/geojson-stream

JSON.parse() is nice because it's very fast and simple to use, but there's that size limitation...

If you want to experiment with an incremental parser, that would be great :)

**Comment by Javier-Rotelli** [comment]
great, i'll be able to work on this in the weekend. 

also, geojson-stream is a wrapper around JSONStream, maybe depending only on the later one is better to support streaming topoJson in the future...

**Comment by gmaclennan** [comment]
one way around this might be to use the programmatic API and pass it parsed json. E.g.

```js
var collect = require('collect-stream')
var geojsonStream = require('geojson-stream')
var mapshaper = require('mapshaper')
var fs = require('fs')

var rs = fs.createReadStream('too_large_to_parse.geojson')

collect(rs.pipe(geojsonStream.parse()), function (err, features) {
  var featureCollection = {
    type: 'FeatureCollection'
    features: features
  }
  mapshaper.applyCommands('-simplify', featureCollection, done)
  function done (err, data) {
    // do something with data (will be stringified JSON)
  }
})
```

**Comment by mbloch** [comment]
I like this suggestion -- definitely worth a try.

**Comment by gmaclennan** [comment]
FYI this did allow me to load and simplify a 300Mb+ GeoJSON file, however I still got allocation errors when the output was large.

The solution is to increase `--max_old_space_size` in node when running your script. e.g.

```sh
node --max_old_space_size=4096 node_modules/.bin/mapshaper -simplify ...
```

> by default v8 has a memory limit of 512MB on 32-bit systems, and 1.4GB on 64-bit systems. The limit can be raised by setting --max_old_space_size to a maximum of ~1024 (~1 GB) (32-bit) and ~4096 (~4GB) (64-bit), but it is recommended that you split your single process into several workers if you are hitting memory limits.

(from the [Node FAQ](https://github.com/nodejs/node/wiki/FAQ#what-is-the-memory-limit-on-a-node-process))

**Comment by mbloch** [comment]
I've found that GeoJSON files tend to balloon up in memory, because they parse to a large number of JS objects (each point location is an array). Very large GeoJSON objects can also cause the v8 garbage collector to become overactive, as it repeatedly scans all of those objects for stuff to collect.

A possible solution is to use the streaming JSON parser to convert GeoJSON coordinates incrementally into the TypedArray objects that mapshaper uses internally. I can look into this when I have some more time to work on mapshaper.

**Comment by gmaclennan** [comment]
That would be great @mbloch. The same streaming processing would be needed for writing / output too. At least it's working for me for now with the solutions above. Thank you for the work on mapshaper, it is incredibly useful!

**Comment by ateucher** [comment]
I don't know if it's helpful in this context, but it looks like there is a proposed standard for incrementally parseable geojson being developed [here](https://github.com/geojson/geojson-text-sequences).

**Comment by mbloch** [comment]
Another option would be to use "Newline-delimited JSON" [ndjson.org](http://ndjson.org/). Mike Bostock has an article about converting GeoJSON to ndjson [here](https://medium.com/@mbostock/command-line-cartography-part-2-c3a82c5c0f3#.j90l8wpkd).

**Comment by mbloch** [comment]
This issue isn't going away -- in fact, more and more people are using very large GeoJSON files. I'll start looking at streaming JSON parsers to replace JSON.parse()... stay tuned.

**Comment by mbloch** [comment]
I just published v0.4.30, which has support for importing large GeoJSON files. The largest file I tested with was a 1.5GB file of ZCTAs, which took about 30 seconds to load.

This doesn't include support for importing large TopoJSON. Also, saving large JSON files is not supported yet...

**Comment by mbloch** [comment]
v0.4.31 supports exporting large GeoJSON files. The largest file I tested was 1.5GB.

If anyone needs support for large TopoJSON files, please file a separate issue.

Closing this one... woot ;)
